fendermint/testing/materializer/src/validation.rs

```diff
@@ -0,0 +1,382 @@
+    references: BTreeMap<SubnetName, HashSet<ResourceHash>>,
+}
+
+impl ValidatingMaterializer {
+    fn network(&self) -> anyhow::Result<TestnetName> {
+        self.network
+            .as_ref()
+            .cloned()
+            .ok_or_else(|| anyhow!("network isn't set"))
+    }
+
+    /// Check that a name is within the subnet. This should trivially be true by construction, but still.
+    fn ensure_contains<T: AsRef<ResourceName> + Debug>(&self, name: &T) -> anyhow::Result<()> {
+        let tn = self.network()?;
+        if !tn.contains(name) {
+            bail!("{tn:?} does not contain {name:?}");
+        }
+        Ok(())
+    }
+
+    /// Ensure we aren't reusing references.
+    fn ensure_unique(
+        &mut self,
+        subnet: &SubnetName,
+        reference: Option<ResourceHash>,
+    ) -> anyhow::Result<()> {
+        if let Some(r) = reference {
+            let rs = self.references.entry(subnet.clone()).or_default();
+            if !rs.insert(r) {
+                bail!("a reference is reused in {subnet:?}");
+            }
+        }
+        Ok(())
+    }
+
+    /// Check that an account has a positive balance on a subnet
+    fn ensure_balance(&self, subnet: &SubnetName, account: &AccountName) -> anyhow::Result<()> {
+        match self.balances.get(subnet) {
+            None => bail!("{subnet:?} has not been created"),
+            Some(bs) => match bs.get(account) {
+                None => bail!("{account:?} has no balance on {subnet:?}"),
+                Some(b) if b.is_zero() => bail!("{account:?} has zero balance on {subnet:?}"),
+                Some(_) => Ok(()),
+            },
+        }
+    }
+
+    /// Check that the subnet has been created already.
+    fn ensure_subnet_exists(&self, subnet: &SubnetName) -> anyhow::Result<()> {
+        if !self.balances.contains_key(subnet) {
+            bail!("{subnet:?} has not been created");
+        }
+        Ok(())
+    }
+
+    /// Move funds of an account from the parent to the child subnet.
+    ///
+    /// Fails if either:
+    /// * the parent doesn't exist
+    /// * the child doesn't exist
+    /// * the account doesn't have the funds
+    fn fund_from_parent(
+        &mut self,
+        subnet: &SubnetName,
+        account: &AccountName,
+        amount: TokenAmount,
+        credit_child: bool,
+    ) -> anyhow::Result<()> {
+        let parent = parent_name(subnet)?;
+        self.ensure_subnet_exists(&parent)?;
+        self.ensure_subnet_exists(subnet)?;
+
+        if amount.is_zero() {
+            return Ok(());
+        }
+
+        self.ensure_balance(&parent, account)?;
+
+        let pbs = self.balances.get_mut(&parent).unwrap();
+        let pb = pbs.get_mut(account).unwrap();
+
+        if *pb < amount {
+            bail!("{account:?} has less than {amount} on {parent:?}, cannot fund {subnet:?}");
+        }
+        *pb = pb.clone().sub(amount.clone());
+
+        if credit_child {
```

Comment:
where will the funds go if this is false?

Reply:
They don't go anywhere in this validator. In reality they get locked up in the `SubnetActor` on the parent. For example currently when you call `join` you say how many tokens you want to put down as collateral, and the `SubnetActor` holds on to it until you are ready to `leave`, or if (in the future) you get slashed. 

Disclaimer I think locking collateral like this isn't great, and I proposed https://github.com/consensus-shipyard/ipc-actors/issues/27 to do it differently.

Reply:
got it!